File: simulation-frontend/js/utils/geo-utils.js
// Berlin Transport Time Machine - Geographic Utilities

const GeoUtils = {
    /**
     * Calculate distance between two points using Haversine formula
     * Returns distance in meters
     */
    calculateDistance(lat1, lng1, lat2, lng2) {
        const R = 6371000; // Earth's radius in meters
        const φ1 = this.toRadians(lat1);
        const φ2 = this.toRadians(lat2);
        const Δφ = this.toRadians(lat2 - lat1);
        const Δλ = this.toRadians(lng2 - lng1);
        
        const a = Math.sin(Δφ/2) * Math.sin(Δφ/2) +
                  Math.cos(φ1) * Math.cos(φ2) *
                  Math.sin(Δλ/2) * Math.sin(Δλ/2);
        const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
        
        return R * c;
    },
    
    /**
     * Calculate bearing between two points
     * Returns bearing in degrees (0-360)
     */
    calculateBearing(lat1, lng1, lat2, lng2) {
        const φ1 = this.toRadians(lat1);
        const φ2 = this.toRadians(lat2);
        const Δλ = this.toRadians(lng2 - lng1);
        
        const y = Math.sin(Δλ) * Math.cos(φ2);
        const x = Math.cos(φ1) * Math.sin(φ2) - 
                  Math.sin(φ1) * Math.cos(φ2) * Math.cos(Δλ);
        
        const θ = Math.atan2(y, x);
        
        return (this.toDegrees(θ) + 360) % 360;
    },
    
    /**
     * Interpolate position between two points
     */
    interpolatePosition(lat1, lng1, lat2, lng2, fraction) {
        const φ1 = this.toRadians(lat1);
        const λ1 = this.toRadians(lng1);
        const φ2 = this.toRadians(lat2);
        const λ2 = this.toRadians(lng2);
        
        // Simple linear interpolation for short distances
        if (this.calculateDistance(lat1, lng1, lat2, lng2) < 1000) {
            return {
                lat: lat1 + (lat2 - lat1) * fraction,
                lng: lng1 + (lng2 - lng1) * fraction
            };
        }
        
        // Great circle interpolation for longer distances
        const d = Math.acos(Math.sin(φ1) * Math.sin(φ2) + 
                           Math.cos(φ1) * Math.cos(φ2) * Math.cos(λ2 - λ1));
        
        if (d === 0) {
            return { lat: lat1, lng: lng1 };
        }
        
        const a = Math.sin((1 - fraction) * d) / Math.sin(d);
        const b = Math.sin(fraction * d) / Math.sin(d);
        
        const x = a * Math.cos(φ1) * Math.cos(λ1) + b * Math.cos(φ2) * Math.cos(λ2);
        const y = a * Math.cos(φ1) * Math.sin(λ1) + b * Math.cos(φ2) * Math.sin(λ2);
        const z = a * Math.sin(φ1) + b * Math.sin(φ2);
        
        const φ3 = Math.atan2(z, Math.sqrt(x * x + y * y));
        const λ3 = Math.atan2(y, x);
        
        return {
            lat: this.toDegrees(φ3),
            lng: this.toDegrees(λ3)
        };
    },
    
    /**
     * Calculate speed in km/h between two positions and times
     */
    calculateSpeed(lat1, lng1, time1, lat2, lng2, time2) {
        const distance = this.calculateDistance(lat1, lng1, lat2, lng2); // meters
        const timeDiff = (time2.getTime() - time1.getTime()) / 1000; // seconds
        
        if (timeDiff <= 0) return 0;
        
        const speedMs = distance / timeDiff;
        return speedMs * 3.6; // Convert m/s to km/h
    },
    
    /**
     * Check if point is within Berlin bounds
     */
    isWithinBerlin(lat, lng) {
        return lat >= CONFIG.MAP.BOUNDS.MIN_LAT && 
               lat <= CONFIG.MAP.BOUNDS.MAX_LAT &&
               lng >= CONFIG.MAP.BOUNDS.MIN_LNG && 
               lng <= CONFIG.MAP.BOUNDS.MAX_LNG;
    },
    
    /**
     * Get bounding box for a set of coordinates
     */
    getBoundingBox(coordinates, padding = 0.01) {
        if (!coordinates || coordinates.length === 0) {
            return {
                minLat: CONFIG.MAP.BOUNDS.MIN_LAT,
                maxLat: CONFIG.MAP.BOUNDS.MAX_LAT,
                minLng: CONFIG.MAP.BOUNDS.MIN_LNG,
                maxLng: CONFIG.MAP.BOUNDS.MAX_LNG
            };
        }
        
        const lats = coordinates.map(coord => coord.lat || coord[0]);
        const lngs = coordinates.map(coord => coord.lng || coord[1]);
        
        return {
            minLat: Math.min(...lats) - padding,
            maxLat: Math.max(...lats) + padding,
            minLng: Math.min(...lngs) - padding,
            maxLng: Math.max(...lngs) + padding
        };
    },
    
    /**
     * Check if point is within bounding box
     */
    isWithinBounds(lat, lng, bounds) {
        return lat >= bounds.minLat && lat <= bounds.maxLat &&
               lng >= bounds.minLng && lng <= bounds.maxLng;
    },
    
    /**
     * Convert coordinates to Berlin local grid system (if needed)
     */
    toBerlinGrid(lat, lng) {
        // This would implement conversion to local coordinate system
        // For now, just return lat/lng
        return { x: lng, y: lat };
    },
    
    /**
     * Create GeoJSON Point feature
     */
    createGeoJSONPoint(lat, lng, properties = {}) {
        return {
            type: 'Feature',
            geometry: {
                type: 'Point',
                coordinates: [lng, lat]
            },
            properties: properties
        };
    },
    
    /**
     * Create GeoJSON LineString feature
     */
    createGeoJSONLineString(coordinates, properties = {}) {
        return {
            type: 'Feature',
            geometry: {
                type: 'LineString',
                coordinates: coordinates.map(coord => [
                    coord.lng || coord[1], 
                    coord.lat || coord[0]
                ])
            },
            properties: properties
        };
    },
    
    /**
     * Simplify line using Douglas-Peucker algorithm
     */
    simplifyLine(points, tolerance = 0.001) {
        if (points.length <= 2) return points;
        
        return this.douglasPeucker(points, tolerance);
    },
    
    /**
     * Douglas-Peucker line simplification
     */
    douglasPeucker(points, tolerance) {
        if (points.length <= 2) return points;
        
        let maxDistance = 0;
        let maxIndex = 0;
        
        for (let i = 1; i < points.length - 1; i++) {
            const distance = this.perpendicularDistance(
                points[i], 
                points[0], 
                points[points.length - 1]
            );
            
            if (distance > maxDistance) {
                maxDistance = distance;
                maxIndex = i;
            }
        }
        
        if (maxDistance > tolerance) {
            const left = this.douglasPeucker(points.slice(0, maxIndex + 1), tolerance);
            const right = this.douglasPeucker(points.slice(maxIndex), tolerance);
            
            return left.slice(0, -1).concat(right);
        } else {
            return [points[0], points[points.length - 1]];
        }
    },
    
    /**
     * Calculate perpendicular distance from point to line
     */
    perpendicularDistance(point, lineStart, lineEnd) {
        const A = lineEnd.lat - lineStart.lat;
        const B = lineStart.lng - lineEnd.lng;
        const C = lineEnd.lng * lineStart.lat - lineStart.lng * lineEnd.lat;
        
        return Math.abs(A * point.lng + B * point.lat + C) / 
               Math.sqrt(A * A + B * B);
    },
    
    /**
     * Snap point to nearest position on line
     */
    snapToLine(point, linePoints) {
        let minDistance = Infinity;
        let closestPoint = null;
        let segmentIndex = -1;
        
        for (let i = 0; i < linePoints.length - 1; i++) {
            const segment = this.closestPointOnSegment(
                point, 
                linePoints[i], 
                linePoints[i + 1]
            );
            
            const distance = this.calculateDistance(
                point.lat, point.lng,
                segment.lat, segment.lng
            );
            
            if (distance < minDistance) {
                minDistance = distance;
                closestPoint = segment;
                segmentIndex = i;
            }
        }
        
        return {
            point: closestPoint,
            distance: minDistance,
            segmentIndex: segmentIndex
        };
    },
    
    /**
     * Find closest point on line segment
     */
    closestPointOnSegment(point, segmentStart, segmentEnd) {
        const A = point.lng - segmentStart.lng;
        const B = point.lat - segmentStart.lat;
        const C = segmentEnd.lng - segmentStart.lng;
        const D = segmentEnd.lat - segmentStart.lat;
        
        const dot = A * C + B * D;
        const lenSq = C * C + D * D;
        
        if (lenSq === 0) {
            return { lat: segmentStart.lat, lng: segmentStart.lng };
        }
        
        const param = dot / lenSq;
        
        if (param < 0) {
            return { lat: segmentStart.lat, lng: segmentStart.lng };
        } else if (param > 1) {
            return { lat: segmentEnd.lat, lng: segmentEnd.lng };
        } else {
            return {
                lat: segmentStart.lat + param * D,
                lng: segmentStart.lng + param * C
            };
        }
    },
    
    /**
     * Utility functions
     */
    toRadians(degrees) {
        return degrees * (Math.PI / 180);
    },
    
    toDegrees(radians) {
        return radians * (180 / Math.PI);
    },
    
    /**
     * Format coordinates for display
     */
    formatCoordinates(lat, lng, precision = 6) {
        return {
            lat: parseFloat(lat.toFixed(precision)),
            lng: parseFloat(lng.toFixed(precision))
        };
    },
    
    /**
     * Validate coordinates
     */
    isValidCoordinate(lat, lng) {
        return typeof lat === 'number' && typeof lng === 'number' &&
               lat >= -90 && lat <= 90 &&
               lng >= -180 && lng <= 180 &&
               !isNaN(lat) && !isNaN(lng);
    }
};

// Export for use in other modules
if (typeof window !== 'undefined') {
    window.GeoUtils = GeoUtils;
}

File: simulation-api/app/utils/geo_utils.py
"""
Geographic calculation utilities for simulation API
"""

import math
from typing import Tuple, List, Dict, Any


def calculate_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """
    Calculate the great circle distance between two points 
    on the earth (specified in decimal degrees)
    Returns distance in meters
    """
    # Convert decimal degrees to radians
    lat1, lon1, lat2, lon2 = map(math.radians, [lat1, lon1, lat2, lon2])
    
    # Haversine formula
    dlat = lat2 - lat1
    dlon = lon2 - lon1
    a = math.sin(dlat/2)**2 + math.cos(lat1) * math.cos(lat2) * math.sin(dlon/2)**2
    c = 2 * math.asin(math.sqrt(a))
    
    # Radius of earth in meters
    r = 6371000
    
    return c * r


def calculate_bearing(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """
    Calculate the bearing between two points
    Returns bearing in degrees (0-360)
    """
    lat1, lon1, lat2, lon2 = map(math.radians, [lat1, lon1, lat2, lon2])
    
    dlon = lon2 - lon1
    
    y = math.sin(dlon) * math.cos(lat2)
    x = math.cos(lat1) * math.sin(lat2) - math.sin(lat1) * math.cos(lat2) * math.cos(dlon)
    
    bearing = math.atan2(y, x)
    bearing = math.degrees(bearing)
    bearing = (bearing + 360) % 360
    
    return bearing


def interpolate_position(
    lat1: float, lon1: float, lat2: float, lon2: float, 
    progress: float
) -> Tuple[float, float]:
    """
    Interpolate position between two points
    progress: 0.0 to 1.0 (0 = first point, 1 = second point)
    """
    lat = lat1 + (lat2 - lat1) * progress
    lon = lon1 + (lon2 - lon1) * progress
    return lat, lon


def get_bounding_box(coordinates: List[Tuple[float, float]], padding: float = 0.01) -> Dict[str, float]:
    """
    Get bounding box for a list of coordinates
    padding: additional padding in degrees
    """
    if not coordinates:
        # Default to Berlin bounds
        return {
            "min_lat": 52.3,
            "max_lat": 52.7,
            "min_lng": 13.0,
            "max_lng": 13.8
        }
    
    lats = [coord[0] for coord in coordinates]
    lngs = [coord[1] for coord in coordinates]
    
    return {
        "min_lat": min(lats) - padding,
        "max_lat": max(lats) + padding,
        "min_lng": min(lngs) - padding,
        "max_lng": max(lngs) + padding
    }


def is_within_berlin_bounds(lat: float, lon: float) -> bool:
    """
    Check if coordinates are within Berlin metropolitan area
    """
    # Berlin bounding box (approximate)
    berlin_bounds = {
        "min_lat": 52.3,
        "max_lat": 52.7,
        "min_lng": 13.0,
        "max_lng": 13.8
    }
    
    return (
        berlin_bounds["min_lat"] <= lat <= berlin_bounds["max_lat"] and
        berlin_bounds["min_lng"] <= lon <= berlin_bounds["max_lng"]
    )


def calculate_speed_kmh(
    lat1: float, lon1: float, time1,
    lat2: float, lon2: float, time2
) -> float:
    """
    Calculate speed in km/h between two positions and timestamps
    """
    distance_m = calculate_distance(lat1, lon1, lat2, lon2)
    time_diff_s = (time2 - time1).total_seconds()
    
    if time_diff_s <= 0:
        return 0.0
    
    speed_ms = distance_m / time_diff_s
    speed_kmh = speed_ms * 3.6
    
    return speed_kmh


def create_geojson_point(lat: float, lon: float, properties: Dict[str, Any] = None) -> Dict[str, Any]:
    """Create a GeoJSON Point feature"""
    return {
        "type": "Feature",
        "geometry": {
            "type": "Point",
            "coordinates": [lon, lat]
        },
        "properties": properties or {}
    }


def create_geojson_linestring(coordinates: List[Tuple[float, float]], properties: Dict[str, Any] = None) -> Dict[str, Any]:
    """Create a GeoJSON LineString feature"""
    return {
        "type": "Feature",
        "geometry": {
            "type": "LineString",
            "coordinates": [[lon, lat] for lat, lon in coordinates]
        },
        "properties": properties or {}
    }